internal/storage/conformance/conformance.go

```diff
@@ -0,0 +1,592 @@
+			resourceType: storage.UnversionedTypeFrom(typeAv1),
+			tenancy: &pbresource.Tenancy{
+				Partition: "default",
+				PeerName:  "local",
+				Namespace: storage.Wildcard,
+			},
+			namePrefix: "",
+			results: []*pbresource.Resource{
+				seedData[0],
+				seedData[1],
+				seedData[2],
+				seedData[5],
+			},
+		},
+		"wildcard tenancy, name prefix": {
+			resourceType: storage.UnversionedTypeFrom(typeAv1),
+			tenancy: &pbresource.Tenancy{
+				Partition: storage.Wildcard,
+				PeerName:  storage.Wildcard,
+				Namespace: storage.Wildcard,
+			},
+			namePrefix: "a",
+			results: []*pbresource.Resource{
+				seedData[0],
+				seedData[1],
+				seedData[3],
+				seedData[5],
+				seedData[6],
+			},
+		},
+	}
+
+	t.Run("List", func(t *testing.T) {
+		backend := opts.NewBackend(t)
+		ctx := testContext(t)
+
+		for _, r := range seedData {
+			_, err := backend.WriteCAS(ctx, r, "")
+			require.NoError(t, err)
+		}
+
+		for desc, tc := range testCases {
+			t.Run(desc, func(t *testing.T) {
+				res, err := backend.List(ctx, tc.resourceType, tc.tenancy, tc.namePrefix)
+				require.NoError(t, err)
+				require.ElementsMatch(t, res, tc.results)
+			})
+		}
+	})
+
+	t.Run("WatchList", func(t *testing.T) {
+		for desc, tc := range testCases {
+			t.Run(fmt.Sprintf("%s - initial snapshot", desc), func(t *testing.T) {
+				backend := opts.NewBackend(t)
+				ctx := testContext(t)
+
+				// Write the seed data before the watch has been established.
+				for _, r := range seedData {
+					_, err := backend.WriteCAS(ctx, r, "")
+					require.NoError(t, err)
+				}
+
+				watch, err := backend.WatchList(ctx, tc.resourceType, tc.tenancy, tc.namePrefix)
+				require.NoError(t, err)
+
+				for i := 0; i < len(tc.results); i++ {
+					ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+					t.Cleanup(cancel)
+
+					event, err := watch.Next(ctx)
+					require.NoError(t, err)
+
+					require.Equal(t, pbresource.WatchEvent_OPERATION_UPSERT, event.Operation)
+					require.Containsf(t, tc.results, event.Resource, "resource not in expected results: %s", event.Resource.Id)
+				}
+			})
+
+			t.Run(fmt.Sprintf("%s - following events", desc), func(t *testing.T) {
+				backend := opts.NewBackend(t)
+				ctx := testContext(t)
+
+				watch, err := backend.WatchList(ctx, tc.resourceType, tc.tenancy, tc.namePrefix)
+				require.NoError(t, err)
+
+				// Write the seed data after the watch has been established.
+				for _, r := range seedData {
+					_, err := backend.WriteCAS(ctx, r, "")
+					require.NoError(t, err)
+				}
+
+				for i := 0; i < len(tc.results); i++ {
+					ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+					t.Cleanup(cancel)
+
+					event, err := watch.Next(ctx)
+					require.NoError(t, err)
+
+					require.Equal(t, pbresource.WatchEvent_OPERATION_UPSERT, event.Operation)
+					require.Containsf(t, tc.results, event.Resource, "resource not in expected results: %s", event.Resource.Id)
```

Comment:
Reminder to myself after [our discussion](https://github.com/hashicorp/consul/pull/16538#discussion_r1134443895): we should try to `Read` the resource here, to make sure reads are sequentially consistent with watches (which isn't currently the case). We should also document the expectation too.